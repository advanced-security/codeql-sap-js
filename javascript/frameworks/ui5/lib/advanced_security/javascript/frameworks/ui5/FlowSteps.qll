import javascript
import advanced_security.javascript.frameworks.ui5.UI5

/** External model to a relevant control property */
class ExternalModelToCustomMetadataPropertyStep extends DataFlow::SharedFlowStep {
  override predicate step(
    DataFlow::Node start, DataFlow::Node end, DataFlow::FlowLabel inLabel,
    DataFlow::FlowLabel outLabel
  ) {
    exists(UI5BindingPath bindingPath |
      bindingPath.getModel() = start and
      end =
        bindingPath
            .getControlDeclaration()
            .getDefinition()
            .getMetadata()
            .getProperty(bindingPath.getPropertyName()) and
      // (
      //   inLabel = "taint" and outLabel = "taint"
      //   or
      //   inLabel = bindingPath.getPath() and outLabel = bindingPath.getPath()
      // )
      if any(UI5BindingPath path/* | path.getModel() = start*/ ).getPath() = inLabel
      then inLabel = bindingPath.getPath() and inLabel = outLabel
      else inLabel = outLabel
    )
  }
}

/*
 * <A a="path1"> a.control.js: {path1: {type:string}}
 * <B b="path1"> b.control.js: {path1: {type:string}}
 */

/** Control metadata property being the intermediate flow node */
class CustomMetadataPropertyReadStep extends DataFlow::SharedFlowStep {
  override predicate step(
    DataFlow::Node start, DataFlow::Node end, DataFlow::FlowLabel inLabel,
    DataFlow::FlowLabel outLabel
  ) {
    exists(PropertyMetadata property |
      /* Writing site -> Control property */
      start = property.getAWrite().getArgument(1) and
      end = property
      or
      /* Control property -> Reading site */
      start = property and
      end = property.getARead()
    ) and
    inLabel = outLabel
  }
}

class LocalModelSetPropertyStep extends DataFlow::SharedFlowStep {
  override predicate step(
    DataFlow::Node start, DataFlow::Node end, DataFlow::FlowLabel inLabel,
    DataFlow::FlowLabel outLabel
  ) {
    exists(
      MethodCallNode setPropertyCall, ModelReference modelRef, CustomController controller,
      InternalModelManifest internalModelManifest
    |
      setPropertyCall.getMethodName() = "setProperty" and
      setPropertyCall.getReceiver().getALocalSource() = modelRef and
      /* We're applying TC + since the `modelRef` can be inside a callback argument. */
      modelRef.asExpr().getEnclosingFunction+() = controller.getAHandler().getFunction() and
      controller.getAModelReference() = modelRef and
      /* `modelRef.getModelName()` can be found in manifest.js */
      internalModelManifest.getName() = modelRef.getModelName() and
      setPropertyCall.getArgument(1) = start and
      modelRef = end and
      /* Any inLabel */
      inLabel = inLabel and
      outLabel =
        modelRef.getModelName() + ">" +
          setPropertyCall.getArgument(0).getALocalSource().asExpr().getStringValue()
    )
  }
}

class LocalModelGetPropertyStep extends DataFlow::SharedFlowStep {
  override predicate step(
    DataFlow::Node start, DataFlow::Node end, DataFlow::FlowLabel inLabel,
    DataFlow::FlowLabel outLabel
  ) {
    exists(
      MethodCallNode getPropertyCall, ModelReference modelRefTo, ModelReference modelRefFrom,
      MethodCallNode setPropertyCall
    |
      setPropertyCall.getMethodName() = "setProperty" and
      setPropertyCall.getReceiver().getALocalSource() = modelRefFrom and
      start = modelRefFrom and
      getPropertyCall.getMethodName() = "getProperty" and
      getPropertyCall.getReceiver().getALocalSource() = modelRefTo and
      inLabel =
        modelRefTo.getModelName() + ">" +
          getPropertyCall.getArgument(0).getALocalSource().asExpr().getStringValue() and
      outLabel = "taint" and
      end = getPropertyCall and
      /*
       * Ensure that getPropertyCall and setPropertyCall are both reading/writing from/to
       * the (1) same property of the (2) same model.
       */

      getPropertyCall.getArgument(0).getALocalSource().asExpr().getStringValue() =
        setPropertyCall.getArgument(0).getALocalSource().asExpr().getStringValue() and
      modelRefFrom.getModelName() = modelRefTo.getModelName()
    )
  }
}

class LocalModelControlMetadataStep extends DataFlow::SharedFlowStep {
  override predicate step(
    DataFlow::Node start, DataFlow::Node end, DataFlow::FlowLabel inLabel,
    DataFlow::FlowLabel outLabel
  ) {
    exists(
      ModelReference modelRef, BindingPath bindingPath, Binding binding, CustomControl control,
      MethodCallNode setPropertyCall
    |
      setPropertyCall.getMethodName() = "setProperty" and
      setPropertyCall.getReceiver().getALocalSource() = modelRef and
      bindingPath = binding.getBindingPath() and
      bindingPath.asString() =
        modelRef.getModelName() + ">" +
          setPropertyCall.getArgument(0).getALocalSource().asExpr().getStringValue() and
      start = modelRef and
      // modelRef.getModelName() = binding.getBindingPath().getModelName() and
      binding.getBindingPath().asString() = inLabel and
      control.getMetadata().getProperty(binding.getBindingTarget().asXmlAttribute().getName()) = end and
      outLabel = "taint"
    )
  }
}

class SetModelToGetModelStep extends DataFlow::SharedFlowStep {
  override predicate step(
    DataFlow::Node start, DataFlow::Node end, DataFlow::FlowLabel inLabel,
    DataFlow::FlowLabel outLabel
  ) {
    exists(UI5Model modelDefinition, ModelReference modelReference |
      modelReference.getResolvedModel() = modelDefinition and
      start = modelDefinition and
      end = modelReference and
      inLabel = inLabel and // any inLabel
      outLabel = outLabel // any outLabel
    )
  }
}

class GetModelToGetPropertyStep extends DataFlow::SharedFlowStep {
  override predicate step(
    DataFlow::Node start, DataFlow::Node end, DataFlow::FlowLabel inLabel,
    DataFlow::FlowLabel outLabel
  ) {
    exists(ModelReference modelReference, MethodCallNode readingMethodCall |
      readingMethodCall = modelReference.getARead() and
      start = modelReference and
      end = readingMethodCall and
      inLabel = inLabel and // any inLabel
      outLabel = "taint"
    )
  }
}