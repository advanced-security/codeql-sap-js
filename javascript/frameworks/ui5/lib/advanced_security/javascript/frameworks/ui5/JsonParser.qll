module JsonParser {
  signature class SourceSig;

  signature module MakeJsonReaderSig<SourceSig Source> {
    class JsonReader extends Source {
      string getJson();
    }
  }

  /**
   * A naive Json parser without error recovery.
   */
  module Make<SourceSig Source, MakeJsonReaderSig<Source> MakeJsonReader> {
    private newtype TJsonToken =
      MkLeftBracketToken(int begin, int end, string value, MakeJsonReader::JsonReader reader) {
        begin = reader.getJson().indexOf("{") and
        begin = end and
        value = "{"
      } or
      MkRightBracketToken(int begin, int end, string value, MakeJsonReader::JsonReader reader) {
        begin = reader.getJson().indexOf("}") and
        begin = end and
        value = "}"
      } or
      MkLeftSquareBracketToken(int begin, int end, string value, MakeJsonReader::JsonReader reader) {
        begin = reader.getJson().indexOf("[") and
        begin = end and
        value = "["
      } or
      MkRightSquareBracketToken(int begin, int end, string value, MakeJsonReader::JsonReader reader) {
        begin = reader.getJson().indexOf("]") and
        begin = end and
        value = "]"
      } or
      MkWhiteSpaceToken(int begin, int end, string value, MakeJsonReader::JsonReader reader) {
        value = reader.getJson().regexpFind("[\\s\\v\\h]", _, begin) and
        begin + value.length() - 1 = end
      } or
      MkCommaToken(int begin, int end, string value, MakeJsonReader::JsonReader reader) {
        begin = reader.getJson().indexOf(",") and
        begin = end and
        value = ","
      } or
      MkColonToken(int begin, int end, string value, MakeJsonReader::JsonReader reader) {
        begin = reader.getJson().indexOf(":") and
        begin = end and
        value = ":"
      } or
      MkNumberToken(int begin, int end, string value, MakeJsonReader::JsonReader reader) {
        value = reader.getJson().regexpFind("-?[1-9]\\d*(\\.\\d+)?((e|E)?(\\+|-)?\\d+)?", _, begin) and
        begin + value.length() - 1 = end
      } or
      MkStringToken(int begin, int end, string value, MakeJsonReader::JsonReader reader) {
        exists(string literal |
          literal = reader.getJson().regexpFind("(?s)\".*?(?<!\\\\)\"", _, begin) and
          // The string without surrounding quotes.
          value = literal.substring(1, literal.length() - 1) and
          begin + literal.length() - 1 = end
        )
      } or
      MkTrueToken(int begin, int end, string value, MakeJsonReader::JsonReader reader) {
        value = reader.getJson().regexpFind("true", _, begin) and
        begin + value.length() - 1 = end
      } or
      MkFalseToken(int begin, int end, string value, MakeJsonReader::JsonReader reader) {
        value = reader.getJson().regexpFind("false", _, begin) and
        begin + value.length() - 1 = end
      } or
      MkNullToken(int begin, int end, string value, MakeJsonReader::JsonReader reader) {
        value = reader.getJson().regexpFind("null", _, begin) and
        begin + value.length() - 1 = end
      }

    private class JsonToken extends TJsonToken {
      int getBegin() {
        this = MkLeftBracketToken(result, _, _, _)
        or
        this = MkRightBracketToken(result, _, _, _)
        or
        this = MkLeftSquareBracketToken(result, _, _, _)
        or
        this = MkRightSquareBracketToken(result, _, _, _)
        or
        this = MkWhiteSpaceToken(result, _, _, _)
        or
        this = MkCommaToken(result, _, _, _)
        or
        this = MkColonToken(result, _, _, _)
        or
        this = MkNumberToken(result, _, _, _)
        or
        this = MkStringToken(result, _, _, _)
        or
        this = MkTrueToken(result, _, _, _)
        or
        this = MkFalseToken(result, _, _, _)
        or
        this = MkNullToken(result, _, _, _)
      }

      int getEnd() {
        this = MkLeftBracketToken(_, result, _, _)
        or
        this = MkRightBracketToken(_, result, _, _)
        or
        this = MkLeftSquareBracketToken(_, result, _, _)
        or
        this = MkRightSquareBracketToken(_, result, _, _)
        or
        this = MkWhiteSpaceToken(_, result, _, _)
        or
        this = MkCommaToken(_, result, _, _)
        or
        this = MkColonToken(_, result, _, _)
        or
        this = MkNumberToken(_, result, _, _)
        or
        this = MkStringToken(_, result, _, _)
        or
        this = MkTrueToken(_, result, _, _)
        or
        this = MkFalseToken(_, result, _, _)
        or
        this = MkNullToken(_, result, _, _)
      }

      string getValue() {
        this = MkLeftBracketToken(_, _, result, _)
        or
        this = MkRightBracketToken(_, _, result, _)
        or
        this = MkLeftSquareBracketToken(_, _, result, _)
        or
        this = MkRightSquareBracketToken(_, _, result, _)
        or
        this = MkWhiteSpaceToken(_, _, result, _)
        or
        this = MkCommaToken(_, _, result, _)
        or
        this = MkColonToken(_, _, result, _)
        or
        this = MkNumberToken(_, _, result, _)
        or
        this = MkStringToken(_, _, result, _)
        or
        this = MkTrueToken(_, _, result, _)
        or
        this = MkFalseToken(_, _, result, _)
        or
        this = MkNullToken(_, _, result, _)
      }

      string toString() {
        this = MkLeftBracketToken(_, _, _, _) and result = "leftbracket"
        or
        this = MkRightBracketToken(_, _, _, _) and result = "rightbracket"
        or
        this = MkLeftSquareBracketToken(_, _, _, _) and result = "leftsquarebracket"
        or
        this = MkRightSquareBracketToken(_, _, _, _) and result = "rightsquarebracket"
        or
        this = MkWhiteSpaceToken(_, _, _, _) and result = "whitespace"
        or
        this = MkCommaToken(_, _, _, _) and result = "comma"
        or
        this = MkColonToken(_, _, _, _) and result = "colon"
        or
        exists(string val | this = MkNumberToken(_, _, val, _) and result = "integer=" + val)
        or
        exists(string val | this = MkStringToken(_, _, val, _) and result = "string=" + val)
        or
        this = MkTrueToken(_, _, _, _) and result = "true"
        or
        this = MkFalseToken(_, _, _, _) and result = "false"
        or
        this = MkNullToken(_, _, _, _) and result = "null"
      }

      MakeJsonReader::JsonReader getReader() {
        this = MkLeftBracketToken(_, _, _, result)
        or
        this = MkRightBracketToken(_, _, _, result)
        or
        this = MkLeftSquareBracketToken(_, _, _, result)
        or
        this = MkRightSquareBracketToken(_, _, _, result)
        or
        this = MkWhiteSpaceToken(_, _, _, result)
        or
        this = MkCommaToken(_, _, _, result)
        or
        this = MkColonToken(_, _, _, result)
        or
        this = MkNumberToken(_, _, _, result)
        or
        this = MkStringToken(_, _, _, result)
        or
        this = MkTrueToken(_, _, _, result)
        or
        this = MkFalseToken(_, _, _, result)
        or
        this = MkNullToken(_, _, _, result)
      }

      JsonToken getNext() {
        result.getBegin() = this.getEnd() + 1 and
        result.getReader() = this.getReader()
      }

      predicate isFirst() {
        not exists(JsonToken other |
          other.getBegin() < this.getBegin() and other.getReader() = this.getReader()
        )
      }
    }

    private class WhiteSpaceToken extends JsonToken, MkWhiteSpaceToken { }

    private class CommaToken extends JsonToken, MkCommaToken { }

    private class ColonToken extends JsonToken, MkColonToken { }

    private class LeftBracketToken extends JsonToken, MkLeftBracketToken { }

    private class RightBracketToken extends JsonToken, MkRightBracketToken { }

    private class LeftSquareBracketToken extends JsonToken, MkLeftSquareBracketToken { }

    private class RightSquareBracketToken extends JsonToken, MkRightSquareBracketToken { }

    private class StringToken extends JsonToken, MkStringToken {
      predicate contains(JsonToken t) {
        this.getReader() = t.getReader() and
        this.getBegin() < t.getBegin() and
        this.getEnd() > t.getEnd()
      }
    }

    private class NumberToken extends JsonToken, MkNumberToken { }

    private class TrueToken extends JsonToken, MkTrueToken { }

    private class FalseToken extends JsonToken, MkFalseToken { }

    private class NullToken extends JsonToken, MkNullToken { }

    private JsonToken getNextSkippingWhitespace(JsonToken t) {
      result = t.getNext() and
      not result instanceof WhiteSpaceToken
      or
      exists(WhiteSpaceToken ws | ws = t.getNext() | result = getNextSkippingWhitespace(ws))
    }

    private newtype TJsonMemberList =
      EmptyMemberList() or
      ConsMemberList(JsonMember head, JsonMemberList tail) {
        exists(JsonToken first, JsonToken last |
          mkJsonMember(first, head, last) and
          if getNextSkippingWhitespace(last) instanceof CommaToken
          then mkJsonMembers(getNextSkippingWhitespace(getNextSkippingWhitespace(last)), tail, _)
          else tail = EmptyMemberList()
        )
      }

    private JsonMember nthMember(JsonMemberList list, int index) {
      exists(JsonMember h | list = ConsMemberList(h, _) | index = 0 and result = h)
      or
      exists(JsonMemberList t | list = ConsMemberList(_, t) |
        index > 0 and result = nthMember(t, index - 1)
      )
    }

    class JsonMemberList extends TJsonMemberList {
      string toString() {
        this = EmptyMemberList() and result = "{}"
        or
        exists(JsonMember head, JsonMemberList tail, string tailStr |
          this = ConsMemberList(head, tail) and
          "{" + tailStr + "}" = tail.toString() and
          if tailStr != ""
          then result = "{" + head.toString() + ", " + tailStr + "}"
          else result = "{" + head.toString() + "}"
        )
      }

      JsonMember getMember(int index) { result = nthMember(this, index) }

      JsonMember getAMember() { result = getMember(_) }
    }

    private newtype TJsonMember =
      MkJsonMember(string key, JsonValue value) {
        exists(StringToken keyToken, ColonToken colonToken, JsonToken firstValueToken |
          colonToken = getNextSkippingWhitespace(keyToken) and
          firstValueToken = getNextSkippingWhitespace(colonToken) and
          key = keyToken.(JsonToken).getValue() and
          mkJsonValue(firstValueToken, value, _)
        )
      }

    class JsonMember extends TJsonMember {
      string toString() {
        exists(string key, JsonValue value |
          this = MkJsonMember(key, value) and
          result = key + " : " + value.toString()
        )
      }

      string getKey() { this = MkJsonMember(result, _) }

      JsonValue getValue() { this = MkJsonMember(_, result) }
    }

    private predicate mkJsonMember(StringToken first, JsonMember member, JsonToken last) {
      getNextSkippingWhitespace(first) instanceof ColonToken and
      exists(JsonValue value |
        mkJsonValue(getNextSkippingWhitespace(getNextSkippingWhitespace(first)), value, last) and
        member = MkJsonMember(first.getValue(), value)
      )
    }

    private predicate mkJsonMembers(JsonToken first, JsonMemberList members, JsonToken last) {
      exists(JsonMember h, JsonToken memberLast | mkJsonMember(first, h, memberLast) |
        not getNextSkippingWhitespace(memberLast) instanceof CommaToken and
        members = ConsMemberList(h, EmptyMemberList()) and
        last = memberLast
      )
      or
      exists(JsonMember h, JsonToken memberLast, JsonMemberList tail |
        mkJsonMember(first, h, memberLast)
      |
        getNextSkippingWhitespace(memberLast) instanceof CommaToken and
        mkJsonMembers(getNextSkippingWhitespace(getNextSkippingWhitespace(memberLast)), tail, last) and
        members = ConsMemberList(h, tail)
      )
    }

    private newtype TJsonValueList =
      EmptyJsonValueList() or
      ConsJsonValueList(JsonValue head, JsonValueList tail) {
        exists(JsonToken first, JsonToken last |
          mkJsonValue(first, head, last) and
          if getNextSkippingWhitespace(last) instanceof CommaToken
          then mkJsonValues(getNextSkippingWhitespace(getNextSkippingWhitespace(last)), tail, _)
          else tail = EmptyJsonValueList()
        )
      }

    private predicate mkJsonValues(JsonToken first, JsonValueList values, JsonToken last) {
      exists(JsonValue h, JsonToken valueLast | mkJsonValue(first, h, valueLast) |
        not getNextSkippingWhitespace(valueLast) instanceof CommaToken and
        values = ConsJsonValueList(h, EmptyJsonValueList()) and
        last = valueLast
      )
      or
      exists(JsonValue h, JsonToken valueLast, JsonValueList tail |
        mkJsonValue(first, h, valueLast)
      |
        getNextSkippingWhitespace(valueLast) instanceof CommaToken and
        mkJsonValues(getNextSkippingWhitespace(getNextSkippingWhitespace(valueLast)), tail, last) and
        values = ConsJsonValueList(h, tail)
      )
    }

    private JsonValue getNthValue(JsonValueList list, int index) {
      exists(JsonValue h | list = ConsJsonValueList(h, _) | index = 0 and result = h)
      or
      exists(JsonValueList t | list = ConsJsonValueList(_, t) |
        index > 0 and result = getNthValue(t, index - 1)
      )
    }

    class JsonValueList extends TJsonValueList {
      string toString() {
        this = EmptyJsonValueList() and result = "[]"
        or
        exists(JsonValue head, JsonValueList tail, string tailStr |
          this = ConsJsonValueList(head, tail) and
          "[" + tailStr + "]" = tail.toString() and
          if tailStr != ""
          then result = "[" + head.toString() + ", " + tailStr + "]"
          else result = "[" + head.toString() + "]"
        )
      }

      JsonValue getValue(int index) { result = getNthValue(this, index) }

      JsonValue getAValue() { result = getValue(_) }
    }

    private newtype TJsonValue =
      MkJsonNumber(float n, JsonToken source) {
        exists(NumberToken t | t.getValue().toFloat() = n and source = t |
          not any(StringToken str).contains(t)
        )
      } or
      MkJsonString(string s, JsonToken source) {
        exists(StringToken t | t.(JsonToken).getValue() = s and t = source)
      } or
      MkJsonObject(JsonMemberList members, JsonToken source) {
        exists(LeftBracketToken l, RightBracketToken r, JsonToken last |
          mkJsonMembers(getNextSkippingWhitespace(l), members, last) and
          getNextSkippingWhitespace(last) = r and
          source = l
        )
        or
        exists(LeftBracketToken l, RightBracketToken r | getNextSkippingWhitespace(l) = r |
          members = EmptyMemberList() and source = l
        )
      } or
      MkJsonArray(JsonValueList values, JsonToken source) {
        exists(LeftSquareBracketToken l, RightSquareBracketToken r, JsonToken last |
          mkJsonValues(getNextSkippingWhitespace(l), values, last) and
          getNextSkippingWhitespace(last) = r and
          source = l
        )
        or
        exists(LeftSquareBracketToken l, RightSquareBracketToken r |
          getNextSkippingWhitespace(l) = r
        |
          values = EmptyJsonValueList() and source = l
        )
      } or
      MkJsonTrue(JsonToken source) {
        exists(TrueToken t | not any(StringToken str).contains(t) and source = t)
      } or
      MkJsonFalse(JsonToken source) {
        exists(FalseToken t | not any(StringToken str).contains(t) and source = t)
      } or
      MkJsonNull(JsonToken source) {
        exists(NullToken t | not any(StringToken str).contains(t) and source = t)
      }

    private predicate mkJsonValue(JsonToken first, JsonValue value, JsonToken last) {
      first instanceof StringToken and
      first = last and
      value = MkJsonString(first.getValue(), first)
      or
      first instanceof NumberToken and
      first = last and
      value = MkJsonNumber(first.getValue().toFloat(), first)
      or
      first instanceof LeftBracketToken and
      (
        exists(JsonMemberList members, JsonToken membersLast |
          mkJsonMembers(getNextSkippingWhitespace(first), members, membersLast) and
          value = MkJsonObject(members, first) and
          last = getNextSkippingWhitespace(membersLast)
        )
        or
        last = getNextSkippingWhitespace(first) and
        value = MkJsonObject(EmptyMemberList(), first)
      ) and
      last instanceof RightBracketToken
      or
      first instanceof LeftSquareBracketToken and
      (
        exists(JsonValueList values, JsonToken valuesLast |
          mkJsonValues(getNextSkippingWhitespace(first), values, valuesLast) and
          value = MkJsonArray(values, first) and
          last = getNextSkippingWhitespace(valuesLast)
        )
        or
        last = getNextSkippingWhitespace(first) and
        value = MkJsonArray(EmptyJsonValueList(), first)
      ) and
      last instanceof RightSquareBracketToken
      or
      first instanceof TrueToken and
      first = last and
      value = MkJsonTrue(first)
      or
      first instanceof FalseToken and
      first = last and
      value = MkJsonFalse(first)
      or
      first instanceof NullToken and
      first = last and
      value = MkJsonNull(first)
    }

    class JsonValue extends TJsonValue {
      string toString() {
        exists(string value |
          this = MkJsonString(value, _) and
          result = "\"" + value + "\""
        )
        or
        exists(float number | this = MkJsonNumber(number, _) | result = number.toString())
        or
        exists(JsonMemberList members | this = MkJsonObject(members, _) |
          result = members.toString()
        )
        or
        exists(JsonValueList values | this = MkJsonArray(values, _) | result = values.toString())
        or
        this = MkJsonTrue(_) and result = "true"
        or
        this = MkJsonFalse(_) and result = "false"
        or
        this = MkJsonNull(_) and result = "null"
      }

      MakeJsonReader::JsonReader getReader() {
        exists(JsonToken token |
          this = MkJsonString(_, token)
          or
          this = MkJsonNumber(_, token)
          or
          this = MkJsonObject(_, token)
          or
          this = MkJsonArray(_, token)
          or
          this = MkJsonTrue(token)
          or
          this = MkJsonFalse(token)
          or
          this = MkJsonNull(token)
        |
          result = token.getReader()
        )
      }

      string getType() {
        this = MkJsonString(_, _) and result = "string"
        or
        this = MkJsonNumber(_, _) and result = "number"
        or
        this = MkJsonObject(_, _) and result = "object"
        or
        this = MkJsonArray(_, _) and result = "array"
        or
        this = MkJsonTrue(_) and result = "true"
        or
        this = MkJsonFalse(_) and result = "false"
        or
        this = MkJsonNull(_) and result = "null"
      }

      string asString() { this = MkJsonString(result, _) }
    }

    class JsonObject extends JsonValue, MkJsonObject {
      JsonMemberList getMembers() { this = MkJsonObject(result, _) }

      JsonMember getMember(int index) { result = getMembers().getMember(index) }

      JsonMember getAMember() { result = getMember(_) }
    }

    class JsonString extends JsonValue, MkJsonString { }

    class JsonNumber extends JsonValue, MkJsonNumber { }

    class JsonArray extends JsonValue, MkJsonArray {
      JsonValueList getValues() { this = MkJsonArray(result, _) }

      JsonValue getValue(int index) { result = getValues().getValue(index) }

      JsonValue getAValue() { result = getValue(_) }
    }

    JsonValue parse(MakeJsonReader::JsonReader reader) {
      result.getReader() = reader and
      exists(JsonToken firstToken |
        firstToken.isFirst() and
        if firstToken instanceof WhiteSpaceToken
        then mkJsonValue(getNextSkippingWhitespace(firstToken), result, _)
        else mkJsonValue(firstToken, result, _)
      )
    }
  }
}
