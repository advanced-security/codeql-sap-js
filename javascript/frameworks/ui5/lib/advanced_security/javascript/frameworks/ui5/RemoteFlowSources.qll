import javascript
import advanced_security.javascript.frameworks.ui5.UI5
import advanced_security.javascript.frameworks.ui5.UI5View
private import semmle.javascript.frameworks.data.internal.ApiGraphModelsExtensions as ApiGraphModelsExtensions

private class DataFromRemoteControlReference extends RemoteFlowSource, MethodCallNode {
  DataFromRemoteControlReference() {
    exists(UI5Control sourceControl, string typeAlias, ControlReference controlReference |
      ApiGraphModelsExtensions::typeModel(typeAlias, sourceControl.getImportPath(), _) and
      ApiGraphModelsExtensions::sourceModel(typeAlias, _, "remote", _) and
      sourceControl.getAReference() = controlReference and
      controlReference.flowsTo(this.getReceiver()) and
      this.getMethodName() = "getValue"
    )
  }

  override string getSourceType() { result = "Data from a remote control" }
}

class LocalModelContentBoundBidirectionallyToSourceControl extends RemoteFlowSource {
  UI5BindingPath bindingPath;
  UI5Control controlDeclaration;

  LocalModelContentBoundBidirectionallyToSourceControl() {
    exists(UI5InternalModel internalModel |
      this = bindingPath.getNode() and
      (
        this instanceof PropWrite and
        internalModel.getArgument(0).getALocalSource().asExpr() =
          this.(PropWrite).getPropertyNameExpr().getParent+()
        or
        this.asExpr() instanceof StringLiteral and
        internalModel.asExpr() = this.asExpr().getParent()
      ) and
      any(UI5View view).getASource() = bindingPath and
      internalModel.(JsonModel).isTwoWayBinding() and
      controlDeclaration = bindingPath.getControlDeclaration()
    )
  }

  override string getSourceType() {
    result = "Local model bidirectionally bound to a input control"
  }

  UI5BindingPath getBindingPath() { result = bindingPath }

  UI5Control getControlDeclaration() { result = controlDeclaration }
}

abstract class UI5ExternalModel extends UI5Model, RemoteFlowSource {
  abstract string getName();
}

/** Model which gains content from an SAP OData service. */
class ODataServiceModel extends UI5ExternalModel {
  string modelName;

  override string getSourceType() { result = "ODataServiceModel" }

  ODataServiceModel() {
    /*
     * e.g. this.getView().setModel(this.getOwnerComponent().getModel("booking_nobatch"))
     */

    exists(MethodCallNode setModelCall, CustomController controller |
      /*
       * 1. This flows from a DF node corresponding to the parent component's model to the `this.setModel` call
       * i.e. Aims to capture something like `this.getOwnerComponent().getModel("someModelName")` as in
       * `this.getView().setModel(this.getOwnerComponent().getModel("someModelName"))`
       */

      modelName = this.getArgument(0).getALocalSource().asExpr().(StringLiteral).getValue() and
      this.getCalleeName() = "getModel" and
      controller.getOwnerComponentRef().flowsTo(this.(MethodCallNode).getReceiver()) and
      this.flowsTo(setModelCall.getArgument(0)) and
      setModelCall.getMethodName() = "setModel" and
      setModelCall.getReceiver() = controller.getAViewReference() and
      /* 2. The component's manifest.json declares the DataSource as being of OData type */
      controller.getOwnerComponent().getExternalModelDef(modelName).getDataSource() instanceof
        ODataDataSourceManifest
    )
    or
    /*
     * A constructor call to sap.ui.model.odata.v2.ODataModel.
     */

    this instanceof NewNode and
    (
      exists(RequiredObject oDataModel |
        oDataModel.flowsTo(this.getCalleeNode()) and
        oDataModel.getDependencyType() = "sap/ui/model/odata/v2/ODataModel"
      )
      or
      this.getCalleeName() = "ODataModel"
    ) and
    modelName = "<no name>"
  }

  override string getName() { result = modelName }
}

private class RouteParameterAccess extends RemoteFlowSource instanceof PropRead {
  override string getSourceType() { result = "RouteParameterAccess" }

  RouteParameterAccess() {
    exists(
      ControllerHandler handler, RouteManifest routeManifest, ParameterNode handlerParameter,
      MethodCallNode getParameterCall
    |
      handler.isAttachedToRoute(routeManifest.getName()) and
      this.asExpr().getEnclosingFunction() = handler.getFunction() and
      handlerParameter = handler.getParameter(0) and
      getParameterCall.getMethodName() = "getParameter" and
      getParameterCall.getReceiver().getALocalSource() = handlerParameter and
      (
        routeManifest.matchesPathString(this.getPropertyName()) and
        this.getBase().getALocalSource() = getParameterCall
        or
        /* TODO: Why does `routeManifest.matchesPathString` not work for propertyName?? */
        this.getBase().(PropRead).getBase().getALocalSource() = getParameterCall
      )
    )
  }
}

private class DisplayEventHandlerParameterAccess extends RemoteFlowSource instanceof PropRead {
  override string getSourceType() { result = "DisplayEventHandlerParameterAccess" }

  DisplayEventHandlerParameterAccess() {
    exists(DisplayEventHandler handler, MethodCallNode getParameterCall |
      getParameterCall.getMethodName() = "getParameter" and
      this.getBase().getALocalSource() = getParameterCall and
      handler.getParameter(0) = getParameterCall.getReceiver().getALocalSource()
    )
  }
}

/**
 * Method calls that fetch a piece of data either from a library control capable of accepting user input, or from a URI parameter.
 */
private class UI5ExtRemoteSource extends RemoteFlowSource {
  UI5ExtRemoteSource() { this = ModelOutput::getASourceNode("remote").asSource() }

  override string getSourceType() {
    result = "Remote flow" // Don't discriminate between UI5-specific remote flows and vanilla ones
  }
}
