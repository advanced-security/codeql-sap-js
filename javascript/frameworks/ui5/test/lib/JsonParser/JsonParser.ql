import javascript
import advanced_security.javascript.frameworks.ui5.JsonParser

string getStringLiteral() {
    result = any(StringLiteral l).getValue()
}

module StringLiteralJsonParser  = JsonParser<getStringLiteral/0>;

from StringLiteralJsonParser::JsonValue val, StringLiteral lit
where val = StringLiteralJsonParser::parse(lit.getValue())
select lit, val, val.getType()