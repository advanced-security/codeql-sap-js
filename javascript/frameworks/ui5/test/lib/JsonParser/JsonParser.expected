| test.js:1:15:1:17 | '3' | 3 | number |
| test.js:2:15:2:18 | '-1' | -1 | number |
| test.js:3:15:3:19 | '1.0' | 1 | number |
| test.js:4:15:4:20 | '-1.0' | -1 | number |
| test.js:5:15:5:22 | '1.0e+1' | 10 | number |
| test.js:6:15:6:23 | '-1.0e+1' | -10 | number |
| test.js:7:15:7:22 | '1.0e-1' | 0.1 | number |
| test.js:9:15:9:29 | '"hello world"' | "hello world" | string |
| test.js:10:15:10:35 | '"hello ... ld\\\\""' | "hello \\"world\\"" | string |
| test.js:11:16:11:33 | '"hello\\nworld\\n"' | "hello\nworld\n" | string |
| test.js:13:16:13:21 | 'true' | true | true |
| test.js:14:16:14:22 | 'false' | false | false |
| test.js:15:16:15:21 | 'null' | null | null |
| test.js:17:16:17:19 | '[]' | [] | array |
| test.js:18:16:18:20 | '[1]' | [1] | array |
| test.js:19:16:19:22 | '[1,2]' | [1, 2] | array |
| test.js:20:16:20:25 | '["1", 2]' | ["1", 2] | array |
| test.js:22:16:22:19 | '{}' | {} | object |
| test.js:23:16:23:24 | '{"a":1}' | {a : 1} | object |
| test.js:24:16:24:31 | '{"a":1,"b":-1}' | {a : 1, b : -1} | object |
| test.js:25:16:25:90 | '{"a":  ... : 2}]}' | {a : {b : 1, c : true}, e : [1, "2", false], f : [{g : 1, h : 2}]} | object |
